lightningd/peer_control.c

```diff
@@ -1230,44 +1287,75 @@ static struct command_result *json_close(struct command *cmd,
 	const jsmntok_t *idtok;
 	struct peer *peer;
 	struct channel *channel COMPILER_WANTS_INIT("gcc 7.3.0 fails, 8.3 OK");
-	unsigned int *timeout;
+	unsigned int *timeout = NULL;
 	bool force = true;
 	bool do_timeout;
+	const u8 *local_shutdown_script = NULL;
+	unsigned int *old_timeout;
+	bool *old_force;
+	bool old_style = false;
 
 	/* For generating help, give new-style. */
 	if (!params || !deprecated_apis) {
 		if (!param(cmd, buffer, params,
 			   p_req("id", param_tok, &idtok),
 			   p_opt_def("unilateraltimeout", param_number,
 				     &timeout, 48 * 3600),
+			   p_opt("destination", param_bitcoin_address,
+				 &local_shutdown_script),
 			   NULL))
 			return command_param_failed();
 		do_timeout = (*timeout != 0);
 	} else if (params->type == JSMN_ARRAY) {
-		const jsmntok_t *tok;
-
+		const jsmntok_t *firsttok, *secondtok;
 		/* Could be new or old style; get as tok. */
-		if (!param(cmd, buffer, params,
+ 		if (!param(cmd, buffer, params,
 			   p_req("id", param_tok, &idtok),
-			   p_opt("unilateraltimeout_or_force", param_tok, &tok),
-			   p_opt("timeout", param_number, &timeout),
+			   p_opt("unilateraltimeout_or_force",
+				 param_tok_timeout_or_force, &firsttok),
+			   p_opt("destination_or_timeout",
+				 param_tok_dest_or_timeout, &secondtok),
 			   NULL))
 			return command_param_failed();
 
-		if (tok) {
+		if (firsttok) {
 			/* old-style force bool? */
-			if (json_to_bool(buffer, tok, &force)) {
-				/* Old default timeout */
-				if (!timeout) {
-					timeout = tal(cmd, unsigned int);
-					*timeout = 30;
-				}
+			if (json_to_bool(buffer, firsttok, &force))
+				old_style = true;
 			/* New-style timeout */
-			} else {
+			else {
 				timeout = tal(cmd, unsigned int);
-				if (!json_to_number(buffer, tok, timeout)) {
+				if (!json_to_number(buffer, firsttok, timeout))
 					return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
 							    "Expected unilerataltimeout to be a number");
+			}
+
+			if (old_style) {
+				timeout = tal(cmd, unsigned int);
+				/* Old default timeout */
+				if (!secondtok)
+					*timeout = 30;
+				else {
+					if (!json_to_number(buffer, secondtok, timeout))
+						return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+								    "(old-style)close: "
+								    "Expected timeout to be a number. "
+								    "This old-style is deprecated!");
+				}
+			} else {
+				if (secondtok) {
+					enum address_parse_result res;
+					res = json_tok_address_scriptpubkey(cmd,
+									    get_chainparams(cmd->ld),
+									    buffer, secondtok,
+									    &local_shutdown_script);
+					if (res == ADDRESS_PARSE_UNRECOGNIZED)
+						return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+								    "Could not parse destination address");
+					else if (res == ADDRESS_PARSE_WRONG_NETWORK)
+						return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+								    "Destination address is not on network %s",
+								    get_chainparams(cmd->ld)->network_name);
 				}
 			}
 		}
```

Comment:
i'd expect an `else if (secondtok)` here. this should be triggered in the case `close id destination`

Reply:
or `close id timeout`?